### ISSUE TITLE
question: can oasgraph be used to generate a graphql api from a loopback 3 swagger?

### ISSUE THREAD

**Issue by tkrugg** [strongloop/oasgraph#95]
**Describe the bug**
Tried this and ended up with multiple:
```
    {
      "type": "InvalidSchemaType",
      "message": "Request / response schema has no (valid) type: {\"type\":\"object\"}",
      "mitigation": "Fall back to type 'GraphQL String'"
    }
```
**To Reproduce**
1. clone https://github.com/strongloop/loopback-example-access-control
2. npm install && npm start. Server starts on :3000
3. run oasgraph http://localhost:3000/explorer/swagger.json`

**Expected behavior**
run with no error

**Comment by Alan-Cha** [comment]
Hi! I know what's going on!

First of all, you should receive two types of errors. 

One that looks like this:

```
    {
      "type": "MissingResponseSchema",
      "message": "Operation 'POST /users/reset-password' has no (valid) response schema.",
      "mitigation": "Ignore operation"
    }
```

... and one that looks like this:

```
    {
      "type": "InvalidSchemaType",
      "message": "Request / response schema has no (valid) type: {\"type\":\"object\"}",
      "mitigation": "Fall back to type 'GraphQL String'"
    }
```

***

The first error is caused by 204 HTTP code operations that do not return responses. This issue has been explored and resolved [here](https://github.com/strongloop/oasgraph/issues/57). To recap, you cannot define a GraphQL object type (the analog for a REST operation) with no fields (the analog for no response). By default, OASGraph will simply ignore these operations, but if you would like to include the operations, you can use the `-f` or `--fillEmptyResponses` option, which will create a placeholder field in the otherwise empty object types. 

*** 

The second is caused by ambiguous JSON Schema [object definitions](https://json-schema.org/understanding-json-schema/reference/object.html). In the schema produced by LoopBack, there are object type definitions that do not have property definitions. In GraphQL, you must clearly define all data types and so, in these cases where the object type is unknown, OASGraph will simply stringify the JSON data. 

You can still access or provide the data but you will have to JSON parse or stringify it first. 

***

TLDR: These aren't technically "errors". These are simply mismatches between REST-like and GraphQL. In general, the world of REST-like is much more expressive than GraphQL. We have provided ways to mitigate these differences. For empty responses, there is the  `-f` or `--fillEmptyResponses` option and for ambiguous object definitions, we have simply stringify the object. 

***

As a side note, I think we can try to improve how we convey this information. Firstly, the way we handle these two issues is inconsistent. For the first error, we ignore the operation but for the second error, we immediately mitigate it. Perhaps we should ignore both and in the `message` property, explain how to mitigate the error. If that is the case, we should also add a new "stringifyAmbiguousObjects" option. @ErikWittern What do you think?

**Comment by jtan4nih** [comment]
Why is this issue closed? Is it fixed?

**Comment by Alan-Cha** [comment]
@jtan4nih This "error" is not a true error because we have provided a mitigation for them, "Fall back to type 'GraphQL String'". 

To summarize what I wrote before, sometimes there can be a mismatch between GraphQL and the OAS. In GraphQL, you must precisely define what objects look like, i.e. describe every property. In the OAS, you can simply define objects as objects: `{"type":"object"}`. When this happens, we choose to stringify the object and store it as a GraphQL string type rather than create a new object type, as it would be impossible to do so without having prior knowledge of the properties. 

The only way to truly fix this "error" is to define all the properties of the object in the OAS. 

The issue is closed because there was never an issue to begin with. The real problem is lacking documentation, which I will try to improve in the next update.

**Comment by GitMastStas** [comment]
How did it go?

**Comment by Alan-Cha** [comment]
@GitMastStas We have improved schema type identification since then but if a schema type cannot be identified, then we produce the following warning:

```
{
    "typeKey": "UNKNOWN_TARGET_TYPE",
    "message": "No GraphQL target type could be identified for schema '{stringified schema}'".
    "mitigation": "The response will be stored in an arbitrary JSON type."
}
```

...which we hope gives the user a better idea of what's going on.
